docs/content/en/docs/Design/simd.md

```diff
@@ -0,0 +1,112 @@
+    As a consequence, the generated code for the snippet above is actually trailed by a (redundant) extract/insert:
+    ```llvm
+    %extracted = tensor.extract %2[%c4] : tensor<32xi16>
+    %inserted = tensor.insert %extracted into %t2[%c4] : tensor<32xi16>
+    ```
+    In real code, this might generate a long series of such extraction/insertion operations, all extracting from the same (due to CSE) tensor and inserting into the same output tensor. Therefore, the `-collapse-insertion-chains` pass searches for such chains over entire tensors and collapses them.
+    It supports not just chains where the indices match perfectly, but any chain where the relative offset is consistent across the tensor, issuing the a rotation to realize the offset (if the offset is zero, the canonicalization will remove the redundant rotation).
+    Note, that in HECO, insertion/extraction is handled differently, as HECO features a `combine` operation modelling not just simple insertions (`combine(%t0#j, %t1)`) but also more complex operations over slices of tensors (`combine(%t0#[i,j], %t1)`).
+    As a result, the equivalent pass in HECO (`-combine-simplify`) instead joins different `combine` operations, and a later fold removes `combines` that replace the entire target tensor.
+    See issue [#512](https://github.com/google/heir/issues/512) for a discussion on why the `combine` operation is a more powerful framework and what would be necessary to port it to HEIR.
+
+* **Applying Rotate-and-Reduce Patterns** (`-rotate-and-reduce -sccp -canonicalize -cse`):
+    The rotate and reduce pattern (see [Representing FHE SIMD Operations](#representing-fhe-simd-operations) for an example) is an important aspect of accelerating SIMD-style operations in FHE, but it does not follow automatically from the batching rewrites applied so far.
+    As a result, the `-rotate-and-reduce` pass needs to search for sequences of arithmetic operations that correspond to the full folding of a tensor,
+    i.e., patterns such as `t[0]+(t[1]+(t[2]+t[3]+(...)))`, which currently uses a backwards search through the IR, but could be achieved more efficiently through a data flow analysis (c.f. issue [#532](https://github.com/google/heir/issues/523)).
+    In HECO, rotate-and-reduce is handled differently, by identifying sequences of compatible operations prior to batching and rewriting them to "n-ary" operations.
+    However, this approach requires non-standard arithmetic operations and is therefore not suitable for use in HEIR.
+    However, there is likely still an opportunity to make the patterns in HEIR more robust/general (e.g., support constant scalar operands in the fold, or support non-full-tensor folds).
+    See issue [#522](https://github.com/google/heir/issues/522) for ideas on how to make the HEIR pattern more robust/more general.
+
+
+### Insert Rotate Pass
+TODO: Write a detailed description of the rotation insertion pass and the associated target slot analysis.
```

Comment:
File an issue and replace this with `TODO(#123):` where 123 is replaced with the issue number

Reply:
I was originally planning to finish writing this up, but I think I'll take the easy way out and do just that ;)